src/MongoDB.Driver/MongoUrl.cs

```diff
@@ -596,7 +596,8 @@ public MongoCredential GetCredential()
             {
                 return MongoCredential.FromComponents(
                     _authenticationMechanism,
-                    _authenticationSource ?? _databaseName,
+                    _authenticationSource,
+                    _databaseName,
```

Comment:
Authentication source and database name had to be split because GSSAPI and X509 must ignore database name and use authentication source only, but they were used clamped in current implementation, which prevented to distinguish them in `FromComponents` method.